src/extension/nodes/CompoundCloudPlane.cpp

```diff
@@ -0,0 +1,479 @@
+    noise_texture->set_bump_strength(4.0);
+
+    material->set_shader_parameter("noise", noise_texture);
+}
+
+void CompoundCloudPlane::update_cloud_texture() {
+    if (!is_inside_tree() || size <= 0)
+        return;
+
+    if (!densities_texture.is_valid()) {
+        densities_texture.instantiate();
+    }
+
+    godot::Ref<godot::Image> image = godot::Image::create(size, size, false, godot::Image::FORMAT_RGBA8);
+    if (!image.is_valid()) {
+        godot::UtilityFunctions::printerr("Failed to create valid image for densities_texture");
+        return;
+    }
+
+    std::lock_guard<std::mutex> lock(density_mutex);
+
+    for (int x = 0; x < size; ++x) {
+        for (int y = 0; y < size; ++y) {
+            godot::Color pixel_color = godot::Color(0, 0, 0, 0);
+            float total_density = 0.0f;
+            int dominant_compound = -1;
+            float max_density = 0.0f;
+
+            for (int i = 0; i < CLOUDS_IN_ONE; ++i) {
+                if (compounds[i] != -1) {
+                    float compound_density = density[x][y][i];
+                    total_density += compound_density;
+                    if (compound_density > max_density) {
+                        max_density = compound_density;
+                        dominant_compound = i;
+                    }
+                }
+            }
+
+            if (total_density > minimum_visible_density && dominant_compound != -1) {
+                pixel_color = compound_colors[dominant_compound];
+                pixel_color.a = godot::Math::clamp(total_density * 2.0f, 0.0f, 1.0f); 
+            }
+
+            image->set_pixel(x, y, pixel_color);
+        }
+    }
+
+    densities_texture->set_image(image);
+    material->set_shader_parameter("densities", densities_texture);
+}
+
+void CompoundCloudPlane::update_shader_parameters() {
+    material->set_shader_parameter("UVOffset", position);
+    material->set_shader_parameter("BrightnessMultiplier", brightness);
+    material->set_shader_parameter("NoiseScale", 14.0f);
+    material->set_shader_parameter("CLOUD_SPEED", 0.013f);
+}
+
+void CompoundCloudPlane::update_position(const godot::Vector2& new_position) {
+    position = new_position;
+    cached_world_position = get_global_transform().origin;
+    update_shader_parameters();
+}
+
+void CompoundCloudPlane::set_brightness(float new_brightness) {
+    brightness = new_brightness;
+    update_shader_parameters();
+}
+
+void CompoundCloudPlane::add_cloud(int compound, int x, int y, float new_density) {
+    if (x < 0 || x >= size || y < 0 || y >= size || size <= 0)
+        return;
+
+    int index = -1;
+    for (int i = 0; i < CLOUDS_IN_ONE; ++i) {
+        if (compounds[i] == compound) {
+            index = i;
+            break;
+        }
+    }
+
+    if (index == -1)
+        return;
+
+    std::lock_guard<std::mutex> lock(density_mutex);
+    density[x][y][index] += new_density;
+    
+    texture_needs_update = true;
+    set_process(true);

```

Comment:
Have you tested the performance impact of constantly calling the enable processing? Most clouds are basically never empty meaning that I think the overhead from enabling / disabling processing is probably more than what is gained by disabling the processing. I think this probably artificially inflates the benchmark numbers. If required I can modify the benchmark to use all compound cloud types in sequence to ensure fairness against the C# implementation that way.